### ISSUE TITLE
NoCoding shouldn't touch entity at all

### ISSUE THREAD

**Issue by jrudolph** [akka/akka#17068]
Originally reported by @RichardBradley as #16998.

/cc @sirthias

**Comment by RichardBradley** [comment]
I had a look at the code to see if I could make PR for this, but the problem is more complicated than I'd hoped.

See [CodingDirectives.scala:95](https://github.com/akka/akka/blob/b75f9063125ee3a0700d1cb34c86a5afc0701888/akka-http/src/main/scala/akka/http/server/directives/CodingDirectives.scala#L95)

The `NoCoding` isn't calling `mapEntity` (although I think it used to??), it's the `mapEntity(StreamUtils.mapEntityError` bit which rewrites the exception which is causing the trouble.

This is closely related to #17048, I think -- we need a way to "mapEntity" without always promoting Strict entities to Chunked.

Any thoughts on how best to fix this? Is this blocked on #17048?

I am currently working around this in user code with a directive like:

```
  def decompressRequest: Directive0 = {
    requestIsNotCompressed |
      decodeRequestWith(Gzip) |
      decodeRequestWith(Deflate)
  }

  def requestIsNotCompressed: Directive0 = {
    requestEntityEmpty | (
      requestEncodedWith(NoCoding.encoding) &
        cancelRejections(classOf[UnsupportedRequestEncodingRejection]))
  }
```

We could rework the `decodeRequest` directive to look more like that and get rid of `NoCoding` altogether. But would that interfere with the apparent "re-chunking" purpose of NoCoding? See #17063

Thanks,


Rich

**Comment by jrudolph** [comment]
Thanks, @RichardBradley. Again a good observation.

We should prevent requests with NoCoding going through that code path. If we need to special case the decodeRequest(With) directive we need to do it, probably in a similar way as you suggest here.

The `decode-max-bytes-per-chunk` with setting was only introduced to prevent a zip bomp in the case for real decoders and we needed a value where we stop decompressing and send out a chunk. This is not the case for NoCoding so it shouldn't be applied here. So, there was no "re-chunking purpose of NoCoding" intended.

**Comment by RichardBradley** [comment]
OK, thanks.

Also, isn't "`StreamUtils.mapEntityError`" too broad here? What stops it from claiming network related stream errors are due to the encoding? (I haven't tested whether it does or not, but it looks like it might.)

**Comment by 2beaucoup** [comment]
BTW: I made an attempt to restructure things regarding (de-)coding a while ago but hadn't got a chance to finish it up. The idea was to remove compression logic from `NoCoding` [1] and implement `Compression` as a trait [2]. I also tried to get rid of `DataMapper` (why do we need it?) and use `transformDataBytes` directly but didn't get the types right on the `HttpMessage` interfaces.

I also tried to make "stacking" of different (de-)coders possible.

The branch is quite outdated but I could try to bring it up to date if you think it's worthwhile.

[1] https://github.com/2beaucoup/akka/commit/b32aeb5ebcc4b433bc8803b0dbcd921da078651b#diff-59bf8a15d3d8984a8e646359c09d9339

[2] https://github.com/2beaucoup/akka/commit/b32aeb5ebcc4b433bc8803b0dbcd921da078651b#diff-5a3daafd30a99e0263b02379bb3abbb2R16

**Comment by jrudolph** [comment]
Thanks, @2beaucoup. I'll have a look at it.

> I also tried to get rid of DataMapper (why do we need it?) and use transformDataBytes directly but didn't get the types right on the HttpMessage interfaces.

Yep, that's the reason for the `DataMapper`. It's impossible to get the types right otherwise. :)

**Comment by 2beaucoup** [comment]
> Yep, that's the reason for the DataMapper. It's impossible to get the types right otherwise. :)

That makes me feel a little better now. ;)

**Comment by jrudolph** [comment]
Fix in #18362 which is very similar to what you proposed, @RichardBradley, but even simpler.
